### ISSUE TITLE
Corrupt timer in `cascade_one`

### ISSUE THREAD

**Issue by yrashk** [h2o/h2o#3243]
I have a case in which I am using libh2o's HTTP client functionality, and it worked fine when I send a moderate amount of requests. However, I noticed that if I send hundreds of the requests at once (sharing the connection pool and socket pools with targets set for all used identical hosts), I get to this assertion failure:

```
cascade_one:281:last_run=1685245794528, wheel=1, slot=7, timer(0x10f116750)={expire_at=0, cb=0x0}
```

I've also noticed that the same issue occurs if I make every request's `on_body` callback sleep a bit, which occurs after perhaps ten requests.

I noticed that `expire_at` is always set to `0` in this case. Adding this code to `cascade_one` seems to alleviate the problem:

```c
 if (entry->expire_at == 0) {
           h2o_linklist_init_anchor(s);
           break;
 }
```

I am generally erring on the side of this being my fault (could it be some memory-management-related issue?). Still, after a couple of fruitless hours, I decided to ask here **if this invariant is indeed possible and `cascade_one` as it is in `master` does not handle it.**

The code I have is not a minimal reproduction of the problem. Hence I am not showing it here, but I can provide as much detail if necessary.

Thank you for your time.

**Comment by kazuho** [comment]
This almost looks like an error in the application side.

In `h2o_timewheel_create`, `ctx->last_run` is initialized to the current time, and in `h2o_timerwheel_link_abs`, we make sure that `expire_at` is no less than `ctx->last_run`. 

Therefore, `entry->expire_at` cannot be zero. Also, the error log shows that `cb` being NULL, which can never be the case for an active timer.

Considering these two observations, I'm inclined to believe that this is an issue in the application clearing an `h2o_timerwheel_entry_t` while it has been registered to the timerwheel.

**Comment by yrashk** [comment]
Just to point out, expire_at may have become zero due to calls to `h2o_timerwheel_init_entry` (which always sets `expire_at` to zero). It is not called directly but through a macro of `h2o_timer_init`. I saw a few cases where `h2o_timer_init` was not followed by `h2o_timer_link`.

I will continue my hunt. I've tried to model my client code off `httpclient.c` to some degree, and since I am not doing anything with timers explicitly, I suspect this is some memory issue. But I wanted to run this by you, because my temporary patch checking `expire_at` for zero has been instrumental in eliminating (masking?) the problem.

**Comment by yrashk** [comment]
I've used Valgrind, and it helped me figure out the issue. This is what Valgrind has alerted me to:

```
==17770== Conditional jump or move depends on uninitialised value(s)
==17770==    at 0x54FF540: h2o_socketpool_register_loop (socketpool.c:282)
```

Turns out, I was registering the sockpool in the event loop before its initialization and registration's behavior is undefined if the sockpool's content is effectively garbage.

So, always initializing prior to registering in the loop is critical.

Perhaps I can send a PR with some documentation that alerts libh2o users to this fact?
